lib/phoenix_component/macro_component.ex

```diff
@@ -0,0 +1,277 @@
+  but in general it is quite simple:
+
+  1. A HTML tag is represented as `{tag, attributes, children}`
+  2. Text is represented as a plain binary
+  3. Attributes are represented as a list of `{key, value}` tuples where
+     the value is an Elixir AST (which can be a plain binary for simple attributes)
+
+  > #### Limitations {: .warning}
+  > The AST is not whitespace preserving, so in cases where you return a modified AST,
+  > the original whitespace between attributes is lost.
+  >
+  > Also, macro components can currently only contain simple HTML. Any interpolation
+  > like `<%= @foo %>` or components inside are not supported.
+
+  ## Example: a compile-time markdown renderer
+
+  Let's say we want to create a macro component that renders markdown as HTML at
+  compile time. First, we need some library that actually converts the markdown to
+  HTML. For this example, we use [`earmark`](https://hex.pm/packages/earmark).
+
+  We start by defining the module for the macro component:
+
+  ```elixir
+  defmodule MyAppWeb.MarkdownComponent do
+    @behaviour Phoenix.Component.MacroComponent
+
+    @impl true
+    def transform({"pre", attrs, children}, _meta) do
+      markdown = Phoenix.Component.MacroComponent.AST.to_string(children)
+      {:ok, html_doc, _} = Earmark.as_html(markdown)
+
+      {"div", attrs, [html_doc]}
+    end
+  end
+  ```
+
+  That's it. Since the div could contain nested elements, for example when using
+  an HTML code block, we need to convert the children to a string first, using the
+  `Phoenix.Component.MacroComponent.AST.to_string/1` function.
+
+  Then, we can simply replace the element's contents with the returned HTML string from
+  Earmark.
+
+  We can now use the macro component inside our HEEx templates:
+
+      defmodule MyAppWeb.ExampleLiveView do
+        use MyAppWeb, :live_view
+
+        def render(assigns) do
+          ~H\"\"\"
+          <pre :type={ColocatedDemoWeb.Markdown} class="prose mt-8">
+          ## Hello World
+
+          This is some markdown!
+
+          ```elixir
+          defmodule Hello do
+            def world do
+              IO.puts "Hello, world!"
+            end
+          end
+          ```
+          </pre>
+          \"\"\"
+        end
+      end
+
+  Note: this example uses the `prose` class from TailwindCSS for styling.
+
+  One trick to prevent issues with extra whitespace is that we use a `<pre>` tag in the LiveView
+  template, which prevents the `Phoenix.LiveView.HTMLFormatter` from indenting the contents, which
+  would mess with the markdown parsing. When rendering, we replace it with a `<div>` tag in the
+  macro component.
+  """
+
+  @type tag :: binary()
+  @type attributes :: %{atom() => term()}
+  @type children :: [heex_ast()]
+  @type heex_ast :: {tag(), attributes(), children()} | binary()
+
+  @callback transform(heex_ast :: heex_ast(), meta :: map()) :: heex_ast()
```

Comment:
We should document what goes in meta.

Reply:
It is good to have a meta just in case we need to add more stuff in the future without breaking compatibility.